[BN] Enable NHWC in OCL

bghimireamd opened the pull request:
* Enable NHWC for Batch norm forward infer
* Initialize the driver and gtest with similar range of values. 
* added BN activation gtest
* Bug fixes on NCHW

bghimireamd committed:
enable nhwc in ocl

bghimireamd committed:
disable nhwc for ocl kernel and only enable ocl kernel for nchw and variant = 2

CAHEK7 commented on src/solver/batchnorm/forward_spatial_multiple.cpp:
```diff
@@ -40,9 +40,42 @@ namespace solver {
+                      problem.GetBnScale().GetType() == miopenFloat;
+
+    // NCHW is Applicable for variant = 2 only
+    if((!(n < 3) &&
+        !((in_nhw < 33554432 && in_cstride > 1024) ||
+          ((n >= 256) && (in_cstride > 60) && bfpmixparm) || ((in_cstride > 512) && bfpmixparm)) &&
+        !(in_cstride <= 512)) ||
+       !((n > 768) && (in_cstride > 150) && bfp32parm))
```
It's barely readable and probably redundant condition.
May I ask you to do the math and simplify it? (at least replace `!(n < 3)` with `(n >= 3)`, but there are more simplifications possible)

CAHEK7 commented on src/solver/batchnorm/backward_spatial_multiple.cpp:
```diff
@@ -38,9 +38,34 @@ namespace solver {
+    unsigned int in_cstride = h * w;
+    unsigned int in_nhw     = n * in_cstride;
+
+    if(!(in_nhw < (32 * 1024 * 1024) && in_cstride > 1024) &&
+       !(in_nhw < (32 * 1024 * 1024) && in_cstride > 512) && !(in_cstride <= 512))
```
It's not that complex as the condition from src/solver/batchnorm/forward_spatial_multiple.cpp, but still can be simplified, since it contains redundant statements.

CAHEK7 replied:
One example of the transformations:
```
X == in_nhw < (32 * 1024 * 1024)
Y == in_cstride > 1024
Z == in_cstride > 512

!(X & Y) & !(X & Z) & !(!Z)
(!X | !Y) & (!X | !Z) & Z
(!X | !Y) & (!X & Z | !Z & Z) // !Z & Z -> false
(!X | !Y) & !X & Z
!X & Z | !Y & !X & Z
!X & Z & (1 | !Y) // (1 | !Y) -> true
!X & Z
```
So it basically means  `return (in_nhw >= (32 * 1024 * 1024)) && (in_cstride > 512);`
Which probably can be simplified more, since `in_nhw` is `n * in_cstride` and we know for sure that `in_cstride` must be greater than 512.

If there are any doubts about those transformations, here is a proof (tested in excel, lol):
Using that fact that if Y is true, then Z must always be true we can even exclude few cases:
|X|Y|Z|old|new|result|
|-|-|-|-|-|-|
|0|0|0|FALSE|FALSE|TRUE|
|1|0|0|FALSE|FALSE|TRUE|
|0|0|1|TRUE|TRUE|TRUE|
|1|0|1|FALSE|FALSE|TRUE|
|0|1|1|TRUE|TRUE|TRUE|
|1|1|1|FALSE|FALSE|TRUE|

CAHEK7 commented on src/solver/batchnorm/backward_spatial_multiple.cpp:
```diff
@@ -38,9 +38,34 @@ namespace solver {
 
+bool BNBwdIsCaseVariant2(const miopen::batchnorm::ProblemDescription& problem)
+{
+    int n, c, h, w;
```
```suggestion
    size_t n, c, h, w;
```

CAHEK7 commented on src/solver/batchnorm/backward_spatial_multiple.cpp:
```diff
@@ -38,9 +38,34 @@ namespace solver {
+    int n, c, h, w;
+    std::tie(n, c, h, w) = tien<4>(problem.GetXDesc().GetLengths());
+
+    unsigned int in_cstride = h * w;
+    unsigned int in_nhw     = n * in_cstride;
```
```suggestion
    size_t in_cstride = h * w;
    size_t in_nhw     = n * in_cstride;
```

CAHEK7 commented on src/solver/batchnorm/backward_spatial_multiple.cpp:
```diff
@@ -38,9 +38,34 @@ namespace solver {
 
+bool BNBwdIsCaseVariant2(const miopen::batchnorm::ProblemDescription& problem)
+{
+    int n, c, h, w;
+    std::tie(n, c, h, w) = tien<4>(problem.GetXDesc().GetLengths());
+
+    unsigned int in_cstride = h * w;
```
But actually you can avoid all the prevois `int`-related comment by the following code:
```suggestion
   auto [n, c, h, w] = tien<4>(problem.GetXDesc().GetLengths());
   auto in_cstride = problem.GetXDesc().GetStrides()[1];
```

CAHEK7 commented on src/solver/batchnorm/forward_spatial_multiple.cpp:
```diff
@@ -40,9 +40,42 @@ namespace solver {
+bool BNFwdTrainIsCaseVariant2(const miopen::batchnorm::ProblemDescription& problem)
+{
+    const auto& xDesc = problem.GetXDesc();
+    int n, c, h, w;
+    std::tie(n, c, h, w)    = tien<4>(xDesc.GetLengths());
+    unsigned int in_cstride = h * w;
+    unsigned int in_nhw     = n * in_cstride;
```
See the comments from src/solver/batchnorm/backward_spatial_multiple.cpp

CAHEK7 commented on src/solver/batchnorm/backward_spatial_multiple.cpp:
```diff
@@ -38,9 +38,34 @@ namespace solver {
+    unsigned int in_cstride = h * w;
+    unsigned int in_nhw     = n * in_cstride;
+
+    if(!(in_nhw < (32 * 1024 * 1024) && in_cstride > 1024) &&
+       !(in_nhw < (32 * 1024 * 1024) && in_cstride > 512) && !(in_cstride <= 512))
+    {
+        return true;
+    }
+    else
+        return false;
+}
```
```suggestion
    return !(in_nhw < (32 * 1024 * 1024) && in_cstride > 1024) &&
       !(in_nhw < (32 * 1024 * 1024) && in_cstride > 512) && !(in_cstride <= 512);
```

bghimireamd committed:
enable nhwc in bn infer, address review comments

bghimireamd committed:
fix clang format

bghimireamd committed:
fix hip tidy issue

bghimireamd committed:
hip cleanups

bghimireamd committed:
fix index for nchw and nhwc for MIOpenBatchNormFwdInferPerAct.cl

xinlipn committed:
Merge branch 'develop' into bg/enable_nhwc_in_ocl

bghimireamd committed:
disable ck bn solver

bghimireamd committed:
bg/enable_nhwc_in_ocl : fix tensor initialization of driver. Make it similar to gtest

xinlipn committed:
3 test cases for backward spatial single and multiple

bghimireamd committed:
fix clang format

bghimireamd committed:
fix type issue

bghimireamd committed:
remove ncwh guard for fwd train multiple

xinlipn committed:
Merge branch 'develop' into sl/bn_bwd

xinlipn committed:
Merge branch 'develop' into sl/bn_bwd

xinlipn committed:
Merge branch 'develop' into sl/bn_bwd

xinlipn committed:
Fix tolerance issue for bwd FP32 test

EnricoDeg committed:
Enable fwd training variant 2 for NHWC

EnricoDeg committed:
Enable support for variant 2 bwd (FP32 works, mixed precision fails)

ex-rzr committed:
Optimize memory accesses in fwd inference for NHWC

bghimireamd committed:
fix merge conflict

bghimireamd committed:
merge with develop

EnricoDeg committed:
Fix timers in batch norm

ex-rzr committed:
Optimize variant 2 of bwd for NHWC

* the kernels now support xlocalsize > 1, usually it's 64;
* the "Final" kernels are launched with one workgroup in y dim;
* variant 2 is enabled by default for NHWC and float because other variants are not optimized yet.

bghimireamd committed:
fix mix bwd var 2

EnricoDeg committed:
Optimize variant 2 of fwd for NHWC

Same strategy as in bwd

bghimireamd committed:
fixed clang format

bghimireamd committed:
place guard around fp32

bghimireamd committed:
clang format

ex-rzr committed:
Fix stash indices for bwd with fp16 and NHWC

EnricoDeg committed:
Fix fwd variant 2 with fp16

EnricoDeg committed:
Use variant 2 as default for fwd with NHWC

EnricoDeg committed:
Fix 3D case for NHWC

bghimireamd committed:
Merge branch 'develop' of github.com:ROCm/MIOpen into bg/enable_nhwc_in_ocl

EnricoDeg committed:
Fix fwd logic

ex-rzr committed:
Add restiction checking for bwd variant 2

* BnBwdTrainingSpatialSingle checks first if BnBwdTrainingSpatialMultiple is applicable (variant 2);
* Change stash indices for NHWC and mixed precision;
* Add restictions for c and h*w for stashing values correctly;

bghimireamd committed:
Merge branch 'bg/bring_stream_hpc_bn_changes' into bg/enable_nhwc_in_ocl

bghimireamd committed:
remove unused macro

bghimireamd committed:
clang format

EnricoDeg committed:
Refactor fwd restriction checking

Consistency with bwd and clearer logic

bghimireamd committed:
fix merge conflict

bghimireamd committed:
fix clang format

bghimireamd committed:
remove unused variable from fwd train ocl bn

xinlipn committed:
[gTest] Integrate Per Activation  (#3498)

* Added miopenBNPerActivation for bn_fwd, bn_infer, bn_fwd_train

* Update CPU Inference and Fwd Train

* Unified interface of batchNormSpatialHostBwdTrain, and batchNormPerActHostBwdTrain

* Update dx_out data type

---------

Co-authored-by: Bibek Ghimire <bghimire@amd.com>

bghimireamd committed:
add comments

BrianHarrisonAMD commented on driver/bn_driver.hpp:
```diff
@@ -195,10 +195,10 @@ int BatchNormDriver<TInput, Tref, TAcc, TScaleBias, TOut>::GetandSetData()
-    {
-        in.GetVector()[i] = prng::gen_canonical<TInput>();
-    }
+    // 0.0 to 2.0 (since unsigned)
+    in.GetTensor().generate([](auto...) {
+        return prng::gen_descreet_unsigned<TInput>(2e-3 /*scale*/, 1000 /*range*/);
+    });
```
Why unsigned?

bghimireamd replied:
good point. will switch to `gen_descreet_signed`

BrianHarrisonAMD commented on driver/bn_driver.hpp:
```diff
@@ -208,20 +208,25 @@ int BatchNormDriver<TInput, Tref, TAcc, TScaleBias, TOut>::GetandSetData()
-            scale.GetVector()[i] = prng::gen_canonical<TInput>();
-            bias.GetVector()[i]  = prng::gen_canonical<TInput>();
-        }
+        // -2.0 to 2.0
+        scale.GetTensor().generate([](auto...) {
+            return prng::gen_descreet_uniform_sign<TScaleBias>(2e-3 /*scale*/, 1000 /*range*/);
+        });
+        bias.GetTensor().generate([](auto...) {
+            return prng::gen_descreet_uniform_sign<TScaleBias>(2e-3 /*scale*/, 1000 /*range*/);
+        });
     }
```
We re-use the same lambda in quite a few places.
Could we declare it once, and re-use it?

bghimireamd replied:
ok will do

BrianHarrisonAMD commented on driver/bn_driver.hpp:
```diff
@@ -208,20 +208,25 @@ int BatchNormDriver<TInput, Tref, TAcc, TScaleBias, TOut>::GetandSetData()
 
-        auto gen_value_emean = [](auto...) { return prng::gen_descreet_unsigned<TAcc>(1e-2, 100); };
-        estMean.InitHostData(estMean.GetTensor().desc.GetElementSize(), true, gen_value_emean);
+        // 0.0 to 1.0
+        estMean.InitHostData(estMean.GetTensor().desc.GetElementSize(), true, [](auto...) {
+            return prng::gen_descreet_uniform_sign<TAcc>(2e-3 /*scale*/, 1000 /*range*/);
+        });
```
Isn't this also -2 to 2?

BrianHarrisonAMD commented on src/batch_norm.cpp:
```diff
@@ -109,14 +104,11 @@ void profileSequence(const Handle& handle, unsigned char select, float* ctime)
-            *ctime = 0.;
-            handle.ResetKernelTime();
             ktime  = handle.GetKernelTime();
             *ctime = ktime;
```
Huh, this would be always 0 before since it reset before getting the kernel time right?

ex-rzr replied:
Yes, that's why the first kernel was ignored in the total duration

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -638,93 +638,122 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
+    gcn_reduce2(&mean, &variance, INHW, lcl_data_x, lcl_data_y, ylid);
 #endif
 
-    barrier(CLK_LOCAL_MEM_FENCE | CLK_GLOBAL_MEM_FENCE);
```
Just to confirm, we believe this isn't needed since the reduce already handles synchronization correctly right?
I think that is the case, but also always a bit worried seeing a barrier removed.

ex-rzr replied:
The kernel does this (correct me if I'm wrong)
1. reads values from global memory `meanvarbuff`
2. reduces them, with synchronization required for local memory
3. writes the results to global memory `meanvarbuff`

I don't see any ways that allows any thread to reach (3) without finishing all reads in (1) and synchronization in (2).

Also, I haven't checked (I don't know how to dump ISA for the kernels in MIOpen), but I suspect that `CLK_GLOBAL_MEM_FENCE` may be extremely slow on MI300.

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -638,93 +638,122 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
 #else
-    commitID = 64;
     local _FLOAT_ACCUM lcl_data_x[MIO_BN_LDSGCN_SIZE];
     local _FLOAT_ACCUM lcl_data_y[MIO_BN_LDSGCN_SIZE];
-    gcn_reduce2(&mean, &variance, FLOAT2ACCUM(INHW), lcl_data_x, lcl_data_y, lid);
+    gcn_reduce2(&mean, &variance, INHW, lcl_data_x, lcl_data_y, ylid);
```
Does this not need to have the same update as the lds_reduce2_2d above with the change to data pointers etc?

ex-rzr replied:
`gcn_reduce2` calculates reduction over the whole workgroup, it can't be used with more than 1 channel per workgroup (1 channel is currently used with NCHW, or  NHWC, fp32 and C = 1).

On RDNA, MIO_BN_GRP1 = 1 (1 channel per workgroup) the new `lds_reduce2_2d` == the old `lds_reduce2`.

BrianHarrisonAMD commented on src/kernels/batchnorm_functions.h:
```diff
@@ -208,6 +208,62 @@
+#if MIOPEN_USE_FPMIX || MIOPEN_USE_BFPMIX
+#if MIO_LAYOUT_NHWC
+    unsigned int index = (ygroupoffset + vindex * 2 + xlid % 2) * ystride +
+                         (xgrp_sz * xgrp_id + xlid / 2 * 2) * xstride;
```
``` xlid / 2 * 2 ``` 
is the intent here to offset by evens?

ex-rzr replied:
We need two 16-bit positions to store one 32-bit value. To make it less restrictive on values of C (and C % xlocalsize, i.e. the size of the last workgroup) , we store even and odd values in different rows (HW dim).

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -638,93 +638,122 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
-        meanvarbuff[varstashindex]  = invVariance; // stash mean
+        // Replicate mean and variance for all y groups because stash == dx_out and
+        // MIOpenBatchNormBwdSpatialDX will read them and rewrite the buffer entirely.
+        storeToStash(
+            mean, meanvarbuff, 2, ygrp_sz * yoffset, ystride, xgrp_sz, xgrp_id, xlid, xstride);
+        storeToStash(invVariance,
+                     meanvarbuff,
+                     3,
+                     ygrp_sz * yoffset,
+                     ystride,
+                     xgrp_sz,
+                     xgrp_id,
+                     xlid,
+                     xstride);
```
Why is vindex 2 & 3? 
I was expecting 0 & 1.

bghimireamd replied:
during storing 0 and 1 is for `MIOpenBatchNormBwdSpatialMeanVariance` and 2 & 3 is for `MIOpenBatchNormBwdSpatialFinalMeanVariance`

BrianHarrisonAMD replied:
Oh looks like it goes up to 5 actually.
Don't we need to increase our index math to account for the fact that we are jumping by 5 elements each time, and to ensure it doesn't overlap?

I guess right now it might be that it's storing these values in the first 12 positions (or 6 if it was FP32) of the meanvarbuff offset by the ygrp_sz, and the assumption is that there will always been enough distance to store that much?

bghimireamd replied:
There should be always space for variant 2 since N* C * H * W value will be large enough to hold at least 12 elements.

BrianHarrisonAMD replied:
Also, looks like we are duplicating the mean & variance for all groups in a loop, but shouldn't it be the same as before where it only writes it once for each group when the commitID == lid? 

Seems like we are duplicating effort here, but I might be missing something that explains why we need this change.

ex-rzr replied:
Actually, I think with the new kernel (that is run with one workgroup in y dim) we can reuse 0, 1 again, because there is no conflict between workgroups as it was before, when multiple groups read and calculated the same values.
That means that we need 4 stashed values, not 6 for bwd.

Let me check to be 100% sure.

ex-rzr replied:
Yes, it's possible now. I added this change to our branch, there are some other changes, it needs to be merged to this one.

fwd now uses 2 values, bwd - 4.

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -733,205 +762,228 @@ MIOpenBatchNormBwdSpatialMeanVariance(const __global _FLOAT* __restrict in,
+                   ygrp_sz);
 #else
-    local _FLOAT_ACCUM lcl_data_x2[MIO_BN_LDSGCN_SIZE];
-    local _FLOAT_ACCUM lcl_data_y2[MIO_BN_LDSGCN_SIZE];
+    local _FLOAT_ACCUM lcl_data_x[MIO_BN_LDSGCN_SIZE];
+    local _FLOAT_ACCUM lcl_data_y[MIO_BN_LDSGCN_SIZE];
     gcn_reduce2(&dscale, &dbias, (_FLOAT_ACCUM)1.0, lcl_data_x2, lcl_data_y2, ylid);
 #endif
```
I think these need similar changes as the lds_reduce2_2d branch above, and also it's still using lcl_data_x2 & lcl_data_y2, but it was renamed to lcl_data_x & lcl_data_y.

BrianHarrisonAMD replied:
We need to test with MIOPEN_USE_AMDGCN true, since it'll still be used on certain devices.

BrianHarrisonAMD replied:
Looking at these defines, I think we need to track if this path gets used. 
It's actually dependent on quite a few defines, and not sure what conditions would use the gcn path.

ex-rzr replied:
Oh, good catch!

The condition must use `MIO_BN_GRP0`, so `lds_reduce2_2d` was always used.

ex-rzr replied:
Fixed in our branch

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -638,93 +638,122 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
+
+    if(xgid >= MIO_BN_C)
+        return;
+
+    _FLOAT_PREC variance = (_FLOAT_PREC)0.;
+    _FLOAT_PREC mean     = (_FLOAT_PREC)0.;
+    _FLOAT_PREC invVariance;
+
+    for(unsigned int yoffset = ylid; yoffset < MIO_BN_NGRPS; yoffset += ygrp_sz)
```
This seems off to me.
Before we were lopping over every group, and then calculating an offset & index for that group.

Now we are incrementing yoffset by the y group size, but then comparing it against the # of groups for terminating the loop.
Is that correct?

ex-rzr replied:
I'm not even sure that the old loop was entirely correct.
Let's assume we have H*W = 10000, NCHW
ygrp_sz = 1024
yngrps = 10
```
for(int gn = 0; gn < yngrps; gn++) // gn = [0..10)
  offset = gn * ygrp_sz + lid; // offset = gn*1024+[0, 1024)
  if(offset < yngrps) // the first iteration is enough to process all yngrps
```
So this loop does more iterations than needed.

Now we iterate over all MIO_BN_NGRPS (remember that in the old kernel yngrps = MIO_BN_NGRPS, in the new kernel yngrps = 1), the loop does as many iterations as needed.

BrianHarrisonAMD replied:
Ah yea the if condition makes it similar. 
It seems odd to me to compare offset to the number of groups, but it was checking the same condition before basically.

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -638,93 +638,122 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
+    unsigned int xstride = MIO_LAYOUT_NHWC ? 1 : MIO_BN_HW;
+    unsigned int ystride = MIO_LAYOUT_NHWC ? MIO_BN_C : 1;
+
+    if(xgid >= MIO_BN_C)
+        return;
```
Since this is variant 2, is the assumption that C will always be >1?

ex-rzr replied:
For fp16, fp32 supports C = 1 and odd C in general

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -638,93 +638,122 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
-    local _FLOAT_ACCUM lcl_data_x[MIO_BN_NGRPS];
-    local _FLOAT_ACCUM lcl_data_y[MIO_BN_NGRPS];
-    lds_reduce2(&mean, &variance, (_FLOAT_ACCUM)1.0, lcl_data_x, lcl_data_y, ylid);
+#if !MIOPEN_USE_AMDGCN || MIO_BN_GRP1 > 1
+    local _FLOAT_ACCUM lcl_data_x[MIO_BN_LDS_SIZE];
+    local _FLOAT_ACCUM lcl_data_y[MIO_BN_LDS_SIZE];
+    lds_reduce2_2d(&mean,
+                   &variance,
+                   (_FLOAT_ACCUM)1.0,
+                   lcl_data_x + xlid * ygrp_sz,
+                   lcl_data_y + xlid * ygrp_sz,
+                   ylid,
+                   ygrp_sz);
```
Similar to above, I think gcn needs the same type of updates, and / or we need to review if gcn is a branch that can be hit. 

If we can't use the gcn path, then I would prefer removing it, and only having the LDS path which is tested and updated.
If we can use the gcn path, then I think we need to update it with similar changes right?

BrianHarrisonAMD replied:
From comment above I think we just need to verify the GCN path.

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -638,93 +638,122 @@ MIOpenBatchNormBwdSpatial(const __global _FLOAT* __restrict x_in,
-    _FLOAT mean            = (_FLOAT)0.;
-    _FLOAT variance        = (_FLOAT)0.;
-    _FLOAT value           = (_FLOAT)0.;
+    _FLOAT_PREC value;
+    _FLOAT_PREC mean     = (_FLOAT_PREC)0.;
+    _FLOAT_PREC variance = (_FLOAT_PREC)0.;
 
     if(ygid < MIO_BN_HW)
     {
-
         for(unsigned int n = 0; n < MIO_BN_N; n++)
         {
-            index = n * MIO_BN_CHW + cidx + ygid;
-            value = *(in + index);
+            index = n * MIO_BN_CHW + ygid * ystride + xgid * xstride;
+            value = FLOAT2FLOATPREC(*(in + index));
             mean += value;
             variance = mad(value, value, variance);
         }
     }
 
-#if !MIOPEN_USE_AMDGCN
-    local _FLOAT_ACCUM lcl_data_x[MIO_BN_NGRPS];
-    local _FLOAT_ACCUM lcl_data_y[MIO_BN_NGRPS];
-    lds_reduce2(&mean, &variance, (_FLOAT_ACCUM)1.0, lcl_data_x, lcl_data_y, ylid);
+#if !MIOPEN_USE_AMDGCN || MIO_BN_GRP1 > 1
+    local _FLOAT_ACCUM lcl_data_x[MIO_BN_LDS_SIZE];
+    local _FLOAT_ACCUM lcl_data_y[MIO_BN_LDS_SIZE];
```
Why do we need to change the lcl data from # of groups to the LDS size?

BrianHarrisonAMD replied:
Actually, was this wrong before?

ex-rzr replied:
Yes, this was incorrect. I spent quite a lot of time to understand what was wrong.

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormBwdSpatial.cl:
```diff
@@ -733,205 +762,228 @@ MIOpenBatchNormBwdSpatialMeanVariance(const __global _FLOAT* __restrict in,
-            ds += *(buff + gammaindex);
-            db += *(buff + betaindex);
-        }
+        dscale +=
+            loadFromStash(buff, 4, ygrp_sz * yoffset, ystride, xgrp_sz, xgrp_id, xlid, xstride);
+        dbias +=
+            loadFromStash(buff, 5, ygrp_sz * yoffset, ystride, xgrp_sz, xgrp_id, xlid, xstride);
     }
 
-#if !MIOPEN_USE_AMDGCN
-    local _FLOAT_ACCUM lcl_data_x2[MIO_BN_NGRPS];
-    local _FLOAT_ACCUM lcl_data_y2[MIO_BN_NGRPS];
-    lds_reduce2(&ds, &db, (_FLOAT_ACCUM)1.0, lcl_data_x2, lcl_data_y2, lid);
+#if !MIOPEN_USE_AMDGCN || MIO_BN_GRP1 > 1
+    local _FLOAT_ACCUM lcl_data_x[MIO_BN_LDS_SIZE];
+    local _FLOAT_ACCUM lcl_data_y[MIO_BN_LDS_SIZE];
+    lds_reduce2_2d(&dscale,
+                   &dbias,
+                   (_FLOAT_ACCUM)1.0,
+                   lcl_data_x + xlid * ygrp_sz,
+                   lcl_data_y + xlid * ygrp_sz,
+                   ylid,
+                   ygrp_sz);
 #else
-    local _FLOAT_ACCUM lcl_data_x2[MIO_BN_LDSGCN_SIZE];
-    local _FLOAT_ACCUM lcl_data_y2[MIO_BN_LDSGCN_SIZE];
-    gcn_reduce2(&ds, &db, (_FLOAT_ACCUM)1.0, lcl_data_x2, lcl_data_y2, lid);
+    local _FLOAT_ACCUM lcl_data_x[MIO_BN_LDSGCN_SIZE];
+    local _FLOAT_ACCUM lcl_data_y[MIO_BN_LDSGCN_SIZE];
+    gcn_reduce2(&dscale, &dbias, (_FLOAT_ACCUM)1.0, lcl_data_x, lcl_data_y, ylid);
 #endif
 
-    if(ygid == 0)
+    if(ylid == 0)
     {
-        delta_scale[xgid] = FLOAT2FLOATPREC(ds);
-        delta_bias[xgid]  = FLOAT2FLOATPREC(db);
+        delta_scale[xgid] = dscale;
+        delta_bias[xgid]  = dbias;
     }
```
Why did this change to the localId instead of the globalId for writing out the values?
Seems like it should be the localId to me, but a bit surprised to see this changed.

ex-rzr replied:
The new Final kernels run with one workgroup in y dim, ygid == ylid

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormFwdTrainSpatial.cl:
```diff
@@ -516,40 +524,47 @@ MIOpenBatchNormFwdTrainSpatialFinalMeanVariance(
+    if(xgid >= MIO_BN_C)
+        return;
+
+    for(unsigned int yoffset = ylid; yoffset < MIO_BN_NGRPS; yoffset += ygrp_sz)
```
Here is another spot I think we need to check the loop conditions.

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormFwdTrainSpatial.cl:
```diff
@@ -559,14 +574,23 @@ MIOpenBatchNormFwdTrainSpatialFinalMeanVariance(
-        meanvarbuff[meanstashindex] = FLOATPREC2FLOAT(mean);        // stash mean
-        meanvarbuff[varstashindex]  = FLOATPREC2FLOAT(invVariance); // stash mean
+
+    for(unsigned int yoffset = ylid; yoffset < MIO_BN_NGRPS; yoffset += ygrp_sz)
+    {
+        storeToStash(
+            mean, meanvarbuff, 2, ygrp_sz * yoffset, ystride, xgrp_sz, xgrp_id, xlid, xstride);
+        storeToStash(invVariance,
+                     meanvarbuff,
+                     3,
+                     ygrp_sz * yoffset,
+                     ystride,
+                     xgrp_sz,
+                     xgrp_id,
+                     xlid,
+                     xstride);
     }
```
Also similar to above.
Is the loop condition correct, and why do we have to store the mean & variance for each group now?

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormFwdTrainSpatial.cl:
```diff
@@ -559,14 +574,23 @@ MIOpenBatchNormFwdTrainSpatialFinalMeanVariance(
     }
 
     // Save mean and calculate and save running mean
-    unsigned int ygid = get_global_id(1);
```
Below there is code saving results assuming xgid == to channel.
Is that a correct assumption for both layouts? 

The x stride makes me think it is, but just want to double check.

ex-rzr replied:
Yes, x - channels, y - HW

BrianHarrisonAMD commented on src/kernels/batchnorm_functions.h:
```diff
@@ -208,6 +208,62 @@
+#if MIOPEN_USE_FPMIX || MIOPEN_USE_BFPMIX
+    // 2 _FLOAT values are used to store 1 _FLOAT_PREC value.
+#if MIO_LAYOUT_NHWC
+    // xgrp_sz values are stpit in two parts: even threads use 2 values at even rows, odd threads -
```
typo on the comment stpit -> split

ex-rzr replied:
Fixed in our branch

BrianHarrisonAMD commented on src/kernels/batchnorm_functions.h:
```diff
@@ -208,6 +208,62 @@
 
 #define UNUSED __attribute__((__unused__))
 
+#if(MIO_BN_VARIANT == 2)
+inline _FLOAT_PREC loadFromStash(const __global _FLOAT* stash,
+                                 unsigned int vindex,
+                                 unsigned int ygroupoffset,
+                                 unsigned int ystride,
+                                 unsigned int xgrp_sz,
+                                 unsigned int xgrp_id,
+                                 unsigned int xlid,
+                                 unsigned int xstride)
+{
```
I think we could unify the load and store index calculation into a single method that is shared.

ex-rzr replied:
Good idea

ex-rzr replied:
Fixed in our branch

BrianHarrisonAMD commented on src/kernels/reduction_functions.h:
```diff
@@ -59,6 +59,30 @@ regLDSreduce(_FLOAT_ACCUM* value, local _FLOAT_ACCUM* data, uint localID, _FLOAT
 
 #endif
 
+static inline void lds_reduce2_2d(_FLOAT_ACCUM* x,
+                                  _FLOAT_ACCUM* y,
+                                  _FLOAT_ACCUM scale,
+                                  local _FLOAT_ACCUM* lcl_data_x,
+                                  local _FLOAT_ACCUM* lcl_data_y,
+                                  unsigned int lid,
+                                  unsigned int size)
+{
+    lcl_data_x[lid] = (_FLOAT_ACCUM)*x;
+    lcl_data_y[lid] = (_FLOAT_ACCUM)*y;
+    barrier(CLK_LOCAL_MEM_FENCE);
+    for(unsigned int red = (size >> 1); red > 0; red >>= 1)
```
Everywhere this is called size == ygrp_sz, and the lcl_data_x / lcl_data_y is initialized to be MIO_BN_LDS_SIZE.

Is the idea that MIO_BN_LDS_SIZE is at least as big as ygrp_sz, and we need this since there is now cases where ygrp_sz is smaller than MIO_BN_LDS_SIZE?

ex-rzr replied:
MIO_BN_LDS_SIZE == ygrp_sz * xgrp_sz always.

Originally, Multiple supported only NCHW with xlocalsize = 1 and ylocalsize = 1024, the naive support of NHWC also used xlocalsize = 1, but this means that all memory accesses are with huge strides, not coalesced (about 10x slower in most cases).
To optimize it we changed workgroup size to shapes like xlocalsize = 64 and ylocalsize = 16. Hence each workgroup needs to do xlocalsize reductions, not 1, because each workgroups processes xlocalsize channels.

There are bank conflicts, hard to say if they affect performance (they can be fixed with padding).

BrianHarrisonAMD commented on src/solver/batchnorm/backward_spatial_multiple.cpp:
```diff
@@ -51,17 +51,50 @@ bool BnBwdTrainingSpatialMultiple::IsApplicable(
+        unsigned int ylocalsize = 1024 / xlocalsize;
+        unsigned int last_ylocalsize =
+            in_cstride % ylocalsize == 0 ? ylocalsize : in_cstride % ylocalsize;
+        if(problem.GetXDesc().GetType() == miopenFloat)
+        {
+            if(last_ylocalsize < stash_values)
+                return false;
+        }
+        else
+        {
+            // Even threads use 2 values at even rows, odd threads - at odd rows.
+            if(c % 2 != 0 || last_ylocalsize < stash_values * 2)
+                return false;
+        }
+    }
+    else
+    {
+        if(!((in_nhw >= static_cast<size_t>(32 * 1024 * 1024) || in_cstride <= 1024) &&
+             in_cstride > 512))
+            return false;
+
+        unsigned int ylocalsize = 1024;
+        unsigned int last_ylocalsize =
+            in_cstride % ylocalsize == 0 ? ylocalsize : in_cstride % ylocalsize;
+        if(last_ylocalsize < stash_values * (problem.GetXDesc().GetType() == miopenFloat ? 1 : 2))
+            return false;
```
Just a nitpick, but it looks like we are checking size compared to stash_values in two different ways here, and it would be nice to check the same way between branches.

BrianHarrisonAMD commented on src/kernels/MIOpenBatchNormFwdTrainSpatial.cl:
```diff
@@ -584,18 +608,24 @@ MIOpenBatchNormFwdTrainSpatialMeanVariance(const __global _FLOAT* __restrict in,
     unsigned int ygrp_sz = get_local_size(1);
     unsigned int index;
-    unsigned int cidx      = xgid * MIO_BN_HW;
-    unsigned int meanindex = cidx + ygrp_sz * ygrp_id;
-    unsigned int varindex  = meanindex + 2;
```
varIndex was always +2 regardless of precision before it looks like.

That's surprising, we will need to make sure everywhere using mvbuff is swapped to the new logic that handles it different depending on the datatype.

ex-rzr replied:
The indices for stashed values were strange:
for bwd they were: (0,2) ; (1,3); (4;6). Why 6, not 5? I don't know.
We changed them to 
(0,1) ; (2,3) ; (4,5) for bwd
(0,1) ; (2,3) for fwd

BrianHarrisonAMD commented on src/solver/batchnorm/backward_spatial_multiple.cpp:
```diff
@@ -51,17 +51,50 @@ bool BnBwdTrainingSpatialMultiple::IsApplicable(
+        }
+        else
+        {
+            // Even threads use 2 values at even rows, odd threads - at odd rows.
+            if(c % 2 != 0 || last_ylocalsize < stash_values * 2)
+                return false;
```
For mixed NHWC multiple can't handle a c that isn't a multiple of 2?

ex-rzr replied:
Yes, because to store one 32-bit intermediate values we use 16-bit values in C dim (for NHWC).
Theoretically, we can store 16-bit halves in different rows (i.e. in HW dim), but that makes indexing even more complicated.

So currently we store as
```
L0,H0,L2,H2,L4,H4...
L1,H1,L3,H3,L5,H5...
...
```
where L and H are low and high 16-bit parts of one 32-bit value.
But we also can store as
```
L0,L1,L2,L3...
H0,H1,H2,H3...
```
There are pros and cons for this approach:
* C can be odd;
* but we need to do two 16-bit loads/stores for each 32-bit value

BrianHarrisonAMD replied:
Does that mean that we have a gap in coverage for FP16 / BF16 mixed if C == 1? 

I thought there was conditions where the single solver couldn't be used due to size limits, and multiple had to be used. 
If that's the case, then I think at a certain size we wouldn't have a solver that could be used when C == 1.

BrianHarrisonAMD commented on src/solver/batchnorm/backward_spatial_single.cpp:
```diff
@@ -40,8 +40,11 @@ namespace solver {
 bool BnBwdTrainingSpatialSingle::IsApplicable(
-    const ExecutionContext&, const miopen::batchnorm::ProblemDescription& bn_problem) const
+    const ExecutionContext& context, const miopen::batchnorm::ProblemDescription& bn_problem) const
 {
+    if(BnBwdTrainingSpatialMultiple{}.IsApplicable(context, bn_problem))
+        return false;
+
```
Did we want to start preferring multiple over single generally?
We used to do the opposite for fwd training at least.

ex-rzr replied:
The issue with the original order is that Single and Multiple need to check exactly the same conditions (and this conditions must ), i.e. Single must know when Multiple can be applied. That's why Single has `return (in_cstride > 1024 && in_nhw < (32 * 1024 * 1024)) ||...` - this is conditions that Multiple also checks.

BrianHarrisonAMD replied:
I think we will need to get @bghimireamd to review since he had changed the IsApplicable quite a bit, and I thought single / multiple had different conditions for support. 

That being said, it does need to prefer one over the other.
If we think multiple (variant 2) is better to prefer, then that seems fine.

BrianHarrisonAMD commented on src/solver/batchnorm/backward_spatial_single.cpp:
```diff
@@ -61,14 +64,7 @@ bool BnBwdTrainingSpatialSingle::IsApplicable(
-
-    unsigned int in_cstride = h * w;
-    unsigned int in_nhw     = n * in_cstride;
-
-    return (in_cstride > 1024 && in_nhw < (32 * 1024 * 1024)) ||
-           (in_cstride > 512 && in_nhw < (32 * 1024 * 1024)) || in_cstride <= 512;
+    return true;
```
Do we not still need these checks?

I thought there was limits to what the single version could support for size.

ex-rzr replied:
These checks are for Multiple, correct me if I'm mistaken.

BrianHarrisonAMD replied:
Do you mean these restrictions were only for multiple?
Or do you mean that this check is for the multiple kernel solution (variant 2)? 

I think we will need to review, but this file is checking the single solver applicability, and I thought there was some limits to what it could support.

BrianHarrisonAMD commented on src/solver/batchnorm/forward_spatial_multiple.cpp:
```diff
@@ -50,7 +48,54 @@ bool BnFwdTrainingSpatialMultiple::IsApplicable(
+        }
+        else
+        {
+            // Even threads use 2 values at even rows, odd threads - at odd rows.
+            if(c % 2 != 0 || last_ylocalsize < stash_values * 2)
+                return false;
```
Training also can't handle c % 2 != 0 NHWC for mixed precision?

EnricoDeg replied:
Yes, the reason is the same as for bwd. Because to store one 32-bit intermediate value, we use 16-bit values in C for NHWC.

BrianHarrisonAMD commented on src/solver/batchnorm/forward_spatial_multiple.cpp:
```diff
@@ -208,9 +213,11 @@ ConvSolution BnFwdTrainingSpatialMultiple::GetSolution(
         result.construction_params.push_back(copy);
 
         copy.kernel_name = kernel.kernel_name + "FinalMeanVariance";
+        copy.g_wk[1]     = kernel.l_wk[1];
```
Why does FinalMeanVariance change the y grid size to match the localsize?

BrianHarrisonAMD replied:
Ah just a single group for FinalMeanVariance?

ex-rzr replied:
Yes, previously the Final kernels run for the full size, basically all groups read and calculated exactly the same final values, BUT they also wrote the finals to their own part of stash (the finals can be stored once because eventually one group would overwrite it with out values as stash == out).

The current approach is to use a single group in y dim (there are still many groups x dim, i.e. for channels, as they are independent). BUT we still need to write the final for each workgroup for the same reason.

This kernel usually takes a small part or the overall duration, but the new approach is faster.

BrianHarrisonAMD commented on src/solver/batchnorm/forward_spatial_single.cpp:
```diff
@@ -40,51 +40,18 @@ namespace solver {
     if(!IsOCLFwdTrainTypeValid(bn_problem))
         return false;
 
-    int n, c, h, w;
-    std::tie(n, c, h, w) = tien<4>(bn_problem.GetXDesc().GetLengths());
-
-    unsigned int in_cstride = h * w;
-    unsigned int in_nhw     = n * in_cstride;
-
-    bool bfpmixparm = false;
-    bool bfp32parm  = true;
-
-    if(bn_problem.GetXDesc().GetType() == miopenHalf &&
-       bn_problem.GetBnScale().GetType() == miopenHalf)
-    {
-        bfp32parm = false;
-    }
-    else if((bn_problem.GetXDesc().GetType() == miopenHalf ||
-             bn_problem.GetXDesc().GetType() == miopenBFloat16) &&
-            bn_problem.GetBnScale().GetType() == miopenFloat)
-    {
-        bfpmixparm = true;
-        bfp32parm  = false;
-    }
-
-    // clang-format off
-    if(!(WORKAROUND_SWDEV_253606 == 0 && n < 3) &&
-        !((in_nhw < 33554432 && in_cstride > 1024) ||
-          ((n >= 256) && (in_cstride > 60) && bfpmixparm) ||
-          ((in_cstride > 512) && bfpmixparm) ||
-          in_cstride <= 512))
-        return false;
-    // clang-format on
-
-    if((n > 768) && (in_cstride > 150) && bfp32parm)
-    {
-        return false;
-    }
-
```
I think we still need some of these checks.
I thought single has limits to the sizes it can handle.

EnricoDeg replied:
The checks are in multiple and then single calls `BnFwdTrainingSpatialMultiple{}.IsApplicable(...)`. In this way, we have all the checks in one place.

BrianHarrisonAMD replied:
I think it'll use multiple if it can, but if it can't, then it falls back to single due to this check:

```C++
if(BnFwdTrainingSpatialMultiple{}.IsApplicable(context, bn_problem))
        return false;
```

If the coverage is identical between single and multiple, then I think this is an issue.
It'll fail the check for multiple, and then potentially allow single afterwards to run.

BrianHarrisonAMD commented on src/solver/batchnorm/forward_spatial_single.cpp:
```diff
@@ -40,51 +40,18 @@ namespace solver {
-    const ExecutionContext&, const miopen::batchnorm::ProblemDescription& bn_problem) const
+    const ExecutionContext& context, const miopen::batchnorm::ProblemDescription& bn_problem) const
 {
+    if(BnFwdTrainingSpatialMultiple{}.IsApplicable(context, bn_problem))
+        return false;
```
Swapping to prefer multiple over single as well here?

EnricoDeg replied:
The previous logic should not be changed. The difference is that all the checks are now in multiple instead of repeating them in single and multiple.

bghimireamd committed:
address review comments partially

BrianHarrisonAMD commented on test/gtest/bn_test_data.hpp:
```diff
@@ -189,12 +199,13 @@ struct BNTestData
 
     void InitTensorsWithRandValue()
     {
-        input.generate(
-            [](auto...) { return prng::gen_descreet_uniform_sign<XDataType>(1e-2, 100); });
+        // 0.0 to 2.0 (since unsigned)
+        input.generate([](auto...) {
+            return prng::gen_descreet_unsigned<XDataType>(2e-3 /*scale*/, 1000 /*range*/);
+        });
```
Similar to driver, any reason this needs to be unsigned?

BrianHarrisonAMD commented on test/gtest/bn_test_data.hpp:
```diff
@@ -253,15 +265,17 @@ struct BNInferTestData : public BNTestData<XDataType, YDataType, TConfig>
 
     void InitTensorsWithRandValue()
     {
-        auto gen_value = [](auto...) {
-            return prng::gen_descreet_uniform_sign<ScaleDataType>(1e-2, 100);
-        };
-        scale.generate(gen_value);
-        shift.generate(gen_value);
-        estMean.generate(gen_value);
-
+        // -2.0 to 2.0
+        scale.generate(
+            [](auto...) { return prng::gen_descreet_uniform_sign<ScaleDataType>(2e-3, 1000); });
+        shift.generate(
+            [](auto...) { return prng::gen_descreet_uniform_sign<BiasDataType>(2e-3, 1000); });
+        estMean.generate(
+            [](auto...) { return prng::gen_descreet_uniform_sign<MeanVarDataType>(2e-3, 1000); });
+
+        // 0.0 to 2.0
         auto gen_var = [](auto...) {
-            return static_cast<MeanVarDataType>(1e-2 * (prng::gen_0_to_B(100) + 1));
+            return static_cast<MeanVarDataType>(2e-3 * (prng::gen_0_to_B(1000) + 1));
         };
         estVariance.generate(gen_var);
```
Same here, why swapping values around?

Probably best to match driver behavior.

ex-rzr committed:
Fix conditions for 2d reduction, gcn_reduce2 was never used enen when possible

EnricoDeg committed:
Unify stash index calculation

ex-rzr committed:
Reuse stash indices

fwd needs 0,1, bwd
This is possible because the new Final kernels run with one
workgroup in y dim, so there are no load/store conflicts between workgroups.

ex-rzr committed:
Use correct out tensor in the driver for bwd per act

bghimireamd committed:
fix bn driver lamda init

bghimireamd committed:
fix merge conflict

bghimireamd committed:
clang format

bghimireamd committed:
add more coverage and few cleanups

bghimireamd committed:
add bn edge test cases

bghimireamd committed:
use double as accumulator and fix minor issues

bghimireamd commented on test/random.hpp:
```diff
@@ -40,5 +40,26 @@ inline T gen_descreet_unsigned(double scale, int32_t range)
+auto uniform_unsigned_initializer(ScaleT scale_arg, RangeT range_arg)
+{
+    return [=](auto&&...) -> T {
+        // uniform sign give balance of both negative and positive values
```
will remove this comment.

bghimireamd committed:
fix bn driver for actiation and cleanups

bghimireamd committed:
fix driver's activation issue

BrianHarrisonAMD reviewed (approved):
LGTM

BrianHarrisonAMD committed:
Merge branch 'develop' into bg/enable_nhwc_in_ocl
